docs/yul.rst

```diff
@@ -1015,3 +1015,117 @@ If you want to use Solidity in stand-alone Yul mode, you activate the optimizer
     solc --strict-assembly --optimize
 
 In Solidity mode, the Yul optimizer is activated together with the regular optimizer.
+
+
+
+Complete ERC20 Example
+======================
+
+::
+
+    object "Token" {
+        code {
+            // Store the creator in slot zero.
+            sstore(0, caller())
+
+            // Deploy the contract
+            datacopy(0, dataoffset("runtime"), datasize("runtime"))
+            return(0, datasize("runtime"))
+        }
+        object "runtime" {
+            code {
+                // Protection against sending Ether
+                if gt(callvalue(), 0) { revert(0, 0) }
+
+                // Dispatcher
+                switch selector()
+                case 0x70a08231 /* "balanceOf(address)" */ {
+                    returnUint(balanceOf(decodeAsAddress(0)))
+                }
+                case 0x18160ddd /* "totalSupply()" */ {
+                    returnUint(totalSupply())
+                }
+                case 0xa9059cbb /* "transfer(address,uint256)" */ {
+                    transfer(decodeAsAddress(0), decodeAsUint(1))
+                }
+                case 0x40c10f19 /* "mint(address,uint256)" */ {
+                    mint(decodeAsAddress(0), decodeAsUint(1))
+                }
+                default {
+                    revert(0, 0)
+                }
+
+                function mint(account, amount) {
```

Comment:
The functions mint, transfer, totalSupply, balanceOf and their internal calls look good to me. However, the address operations need to be examined by someone who know YUL better.

What about the rest of the standard functions, i.e., transferFrom, approve and allowance?

Reply:
As far as I know, transferFrom, approve and allowance are optional in ERC20, and I think we should keep this example minimal.

Reply:
According to https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20.md, they (transferFrom, approve and allowance) are not optional.

Reply:
You are right - sorry about that! If we want to implement it, we need another data structure and we will have to apply a hash.

Reply:
You can also rename it to "BasicToken" to get over the issue.